### PR TITLE
Np 48254 persist nvi candidates with unverified creators

### PR DESCRIPTION
**Pull request by LarsV123**

### CODE REVIEW

File: index-handlers/src/test/java/no/sikt/nva/nvi/index/IndexDocumentHandlerTest.java [hunk 1]
```diff
@@ -183,6 +185,24 @@ void shouldNotFailWhenCandidateIsMissingChannelTypeOrChannelId() {
         assertEquals(expectedIndexDocument, actualIndexDocument);
     }
 
+    @Test
+    void shouldNotFailWhenCreatorIsUnverified() {
+        var institutionId = randomUri();
```
Comment (monaullah) [inline index-handlers/src/test/java/no/sikt/nva/nvi/index/IndexDocumentHandlerTest.java:190]: Maybe 
```
var request = randomUpsertRequestBuilder()
                          .withUnverifiedCreators(
                              List.of(new UnverifiedNviCreator(randomString(), List.of(randomUri()))))
                          .build();
        Candidate.upsert(request, candidateRepository, periodRepository);
        var candidate = Candidate.fetchByPublicationId(request::publicationId, candidateRepository, periodRepository);
```

To avoid using dao when testing index document generation?

File: index-handlers/src/main/java/no/sikt/nva/nvi/index/utils/NviCandidateIndexDocumentGenerator.java [hunk 2]
```diff
@@ -322,21 +324,43 @@ private String extractAssignee(Approval approval) {
...
+                                                                                  Candidate candidate) {
+        return candidate.getPublicationDetails()
+                        .getUnverifiedCreators()
+                        .stream()
+                        .filter(creator -> creator.name()
+                                                  .equals(extractName(contributor.at(JSON_PTR_IDENTITY))))
+                        .findFirst();
+    }
+
+    // FIXME
```
Comment (monaullah) [inline index-handlers/src/main/java/no/sikt/nva/nvi/index/utils/NviCandidateIndexDocumentGenerator.java:352]: Fix what?

File: event-handlers/src/test/java/no/sikt/nva/nvi/events/persist/UpsertNviCandidateHandlerTest.java [hunk 3]
```diff
@@ -307,12 +369,17 @@ private DbCandidate getExpectedCandidate(NviCandidate evaluatedNviCandidate) {
...
+        var verifiedCreators = evaluatedNviCandidate.creators()
+                                  .entrySet()
+                                  .stream()
+                                  .map(entry -> new DbCreator(entry.getKey(), entry.getValue()))
+                                  .toList();
+        var unverifiedCreators = evaluatedNviCandidate.unverifiedCreators()
+                                        .stream()
+                                        .map(UnverifiedNviCreator::toDao)
+                                        .toList();
+        return union(verifiedCreators, unverifiedCreators);
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply event-handlers/src/test/java/no/sikt/nva/nvi/events/persist/UpsertNviCandidateHandlerTest.java:382]: Yes, but couldn't get it working due to type inference issues. This works though, thanks!

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/model/NviCandidate.java [hunk 4]
```diff
@@ -22,7 +23,7 @@ public record NviCandidate(URI publicationId,
                            InstanceType instanceType,
                            @JsonProperty("publicationDate") PublicationDate date,
                            List<NviCreator> nviCreators,
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply event-handlers/src/main/java/no/sikt/nva/nvi/events/model/NviCandidate.java:25]: Agreed, this should be renamed and some models merged. I can't find a clean way to do it though, so I'll do it in a separate PR later.

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/UnverifiedNviCreator.java [hunk 5]
```diff
@@ -1,20 +1,26 @@
...
+    @Override
+    public DbUnverifiedCreator toDao() {
+        return new DbUnverifiedCreator(name, affiliations);
+    }
+
     public static final class Builder {
 
         private String name;
-        private List<URI> nviAffiliations = Collections.emptyList();
+        private List<URI> affiliations = Collections.emptyList();
```
[the remark this answers is not in the document]
Comment (monaullah) [reply nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/UnverifiedNviCreator.java:23]: I prefer explicit, but you decide 🙏

[the remark this answers is not in the document]
Comment (monaullah) [reply nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/UnverifiedNviCreator.java:23]: Since it's a `List` of `URI`s, and not `NviAffiliation`s

**Review by monaullah — APPROVED** [review]

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/db/CandidateDao.java [hunk 6]
```diff
@@ -580,19 +589,40 @@ public DbPublicationDate build() {
         }
     }
 
+    // Note that this has defaultImpl = DbCreator.class to handle old data without type information
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply nvi-commons/src/main/java/no/sikt/nva/nvi/common/db/CandidateDao.java:592]: It could be removed in the future (once all existing data is migrated), but it could be left just to be defensive. It's a reasonable default assumption in the unlikely event that type info is missing.

The rest of the code should remain in any case, but I can add a FIXME comment saying that this property can be removed later? `@Deprecated` is a bit noisy (in my IDE at least).

File: event-handlers/src/test/java/no/sikt/nva/nvi/events/persist/UpsertNviCandidateHandlerTest.java [hunk 7]
```diff
@@ -185,6 +187,65 @@ void shouldNotResetApprovalsWhenUpdatingFieldsNotEffectingApprovals() {
         assertThat(updatedApproval, is(equalTo(approval)));
     }
 
+    @Test
+    void shouldSaveNewNviCandidateWithOnlyUnverifiedCreatorWhenCandidateDoesNotExist() {
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply event-handlers/src/test/java/no/sikt/nva/nvi/events/persist/UpsertNviCandidateHandlerTest.java:191]: Good point! I think the phrasing is inherited from existing tests that were copied and modified, and the `DoesNotExist` part just means it is a new candidate. Other tests specifically apply to updating existing DB entries. It's not really relevant here though.

I think we can shorten this to just `shouldSaveNewNviCandidateWithOnlyUnverifiedCreators`  and `shouldSaveNewNviCandidateWithBothVerifiedAndUnverifiedCreators`

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/cristin/CristinMapper.java [hunk 8]
```diff
@@ -115,9 +116,7 @@ public DbCandidate toDbCandidate(CristinNviReport cristinNviReport) {
                    .basePoints(extractBasePoints(cristinNviReport))
                    .collaborationFactor(extractCollaborationFactor(cristinNviReport))
                    .internationalCollaboration(isInternationalCollaboration(cristinNviReport))
-                   .creators(extractCreators(cristinNviReport))
-                   //                   .creatorCount()
-                   //                   .creatorShareCount()
+                          .creators(extractCreators(cristinNviReport))
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply event-handlers/src/main/java/no/sikt/nva/nvi/events/cristin/CristinMapper.java:119]: Yes, you're probably right. I'd rather not touch it right now if I can avoid it though, since it appears to work...

File: index-handlers/src/test/java/no/sikt/nva/nvi/index/IndexDocumentHandlerTest.java [hunk 9]
```diff
@@ -183,6 +185,26 @@ void shouldNotFailWhenCandidateIsMissingChannelTypeOrChannelId() {
...
+                                                  .build();
+        Candidate.upsert(request, candidateRepository, periodRepository);
+        var candidate = Candidate.fetchByPublicationId(request::publicationId, candidateRepository, periodRepository);
+        var expectedIndexDocument = setupExistingResourceInS3AndGenerateExpectedDocument(
+            candidate).indexDocument();
+
+        mockUriRetrieverOrgResponse(candidate);
+        handler.handleRequest(createEvent(candidate.getIdentifier()), CONTEXT);
+        var actualIndexDocument = parseJson(s3Writer.getFile(createPath(candidate))).indexDocument();
+        assertEquals(expectedIndexDocument, actualIndexDocument);
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply index-handlers/src/test/java/no/sikt/nva/nvi/index/IndexDocumentHandlerTest.java:205]: This handler also appears to swallow exceptions. Is there a reasonable way to test failures directly that I'm missing?

File: buildSrc/src/main/groovy/nva.nvi.rootplugin.gradle [hunk 10]
```diff
@@ -27,6 +27,28 @@ reporting {
...
+            subprojects.each { subproject ->
+                def subprojectReportDir = subproject.layout.buildDirectory.dir(reportDirPath).get().asFile
+                logger.quiet("\t Coverage report for subproject: file://${subprojectReportDir}/index.html (${subproject.name})")
+            }
+            def reportDir = layout.buildDirectory.dir(reportDirPath).get().asFile
+            logger.quiet("Combined coverage report: file://${reportDir}/index.html")
+        }
+    }
+}
+
```
Comment (LarsV123) [inline buildSrc/src/main/groovy/nva.nvi.rootplugin.gradle:51]: This is purely a util task to make local dev easier, could perhaps be moved elsewhere

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/CreatorVerificationUtil.java
Comment (LarsV123) [inline event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/CreatorVerificationUtil.java:1]: This file has been moved and partially rewritten to limit use of JsonNode

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/PointService.java [hunk 11]
```diff
@@ -49,13 +50,14 @@ public PointService(OrganizationRetriever organizationRetriever) {
     }
 
     public PointCalculation calculatePoints(JsonNode publication,
-                                            List<VerifiedNviCreator> nviCreators) {
+                                            List<VerifiedNviCreator> verifiedNviCreators, List<UnverifiedNviCreator> unverifiedNviCreators) {
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/PointService.java:53]: Nope, will replace it!

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/PointService.java [hunk 12]
```diff
@@ -209,6 +211,8 @@ private Integer countVerifiedTopLevelAffiliationsPerCreator(List<JsonNode> creat
                    .reduce(0, Integer::sum);
     }
 
+    // FIXME: Does this actually need to call the organization retriever?
+    // This method is only called from EvaluatorService, which has already done the same check.
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/PointService.java:215]: Done: https://sikt.atlassian.net/browse/NP-48364

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/calculator/PointCalculator.java [hunk 13]
```diff
@@ -98,49 +117,75 @@ private static Stream<CreatorAffiliationPoints> calculatePointsForAffiliation(En
...
+    }
+
+    private BigDecimal getInstitutionContributorFraction(Long institutionCreatorCount) {
+        if (institutionCreatorCount == 0) {
+            return ZERO.setScale(SCALE, ROUNDING_MODE);
+        }
+        return BigDecimal
+                   .valueOf(institutionCreatorCount)
+                   .divide(new BigDecimal(creatorShareCount), MATH_CONTEXT)
+                   .setScale(SCALE, ROUNDING_MODE);
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/calculator/PointCalculator.java]: Like this?

```
    private BigDecimal getInstitutionContributorFraction(Long institutionCreatorCount) {
        var value = isNullOrZero(institutionCreatorCount) ? ZERO : BigDecimal
                                                                       .valueOf(institutionCreatorCount)
                                                                       .divide(new BigDecimal(creatorShareCount),
                                                                               MATH_CONTEXT);
        return value.setScale(SCALE, ROUNDING_MODE);
    }

    private boolean isNullOrZero(Long creatorShareCount) {
        return creatorShareCount == null || creatorShareCount == 0;
    }
```

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/Candidate.java [hunk 14]
```diff
@@ -546,19 +573,30 @@ private static DbApprovalStatus mapToApproval(URI institutionId) {
...
     private static DbCandidate mapToCandidate(UpsertCandidateRequest request) {
-        var creators = getAllCreators(request).stream()
-                                              .map(NviCreatorType::toDao)
-                                              .toList();
-        return DbCandidate.builder()
+        Stream<DbCreatorType> verifiedCreators = request
+                                                     .verifiedCreators()
+                                                     .stream()
+                                                     .map(VerifiedNviCreatorDto::toDao);
+        Stream<DbCreatorType> unverifiedCreators = request
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/Candidate.java:580]: We can, but then we need an explicit cast to `DbCreatorType` (for obscure reasons). Records might be a bad fit for this use-case.

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/Candidate.java [hunk 15]
```diff
@@ -584,10 +622,17 @@ private static DbPublicationDate mapToPublicationDate(PublicationDate publicatio
...
-    private static List<DbCreatorType> mapToDbCreators(List<NviCreatorType> creators) {
-        return creators.stream()
-                       .map(NviCreatorType::toDao)
-                       .toList();
+    private static List<DbCreatorType> mapToDbCreators(List<VerifiedNviCreatorDto> verifiedNviCreators,
+                                                       List<UnverifiedNviCreatorDto> unverifiedNviCreators) {
+        Stream<DbCreatorType> verifiedCreators = verifiedNviCreators
+                                                     .stream()
+                                                     .map(VerifiedNviCreatorDto::toDao);
+        Stream<DbCreatorType> unverifiedCreators = unverifiedNviCreators
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/Candidate.java]: Same here, this needs an explicit cast to work. We can re-use this method at least though, I see I've duplicated the logic.

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/calculator/CreatorVerificationUtil.java [hunk 16]
```diff
@@ -57,56 +47,60 @@ public CreatorVerificationUtil(AuthorizedBackendUriRetriever authorizedBackendUr
...
 
-    public List<UnverifiedNviCreator> getUnverifiedCreatorsWithNviInstitutions(JsonNode body) {
-        return getJsonNodeStream(body, JSON_PTR_CONTRIBUTOR)
+    public List<UnverifiedNviCreator> getUnverifiedCreatorsWithNviInstitutions(List<ContributorDto> contributors) {
+        return contributors
+                   .stream()
                    .filter(not(CreatorVerificationUtil::isVerified))
                    .filter(CreatorVerificationUtil::isCreator)
-                   .filter(CreatorVerificationUtil::hasName)
+                   .filter(not(contributor -> isBlank(contributor.name())))
```
[the remark this answers is not in the document]
Comment (LarsV123) [reply event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/calculator/CreatorVerificationUtil.java:72]: True, tried now. Needed some refactoring to make it work though.

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/calculator/CreatorVerificationUtil.java [hunk 17]
```diff
@@ -47,56 +51,62 @@ public CreatorVerificationUtil(AuthorizedBackendUriRetriever authorizedBackendUr
...
-    private static boolean isVerified(ContributorDto contributor) {
-        return VERIFIED.equals(contributor.verificationStatus());
+    private static boolean isValidContributor(ContributorDto contributorDto) {
+        return isVerifiedContributor(contributorDto) || isNamedContributor(contributorDto);
+    }
+
+    private static boolean isVerifiedContributor(ContributorDto contributorDto) {
+        var status = contributorDto.verificationStatus();
+        var id = contributorDto.id();
+        return VERIFIED.equals(status) && id != null;
```
Comment (brinxmat) [inline event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/calculator/CreatorVerificationUtil.java:99]: Objects::nonNull

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/calculator/PointCalculator.java [hunk 18]
```diff
@@ -141,13 +142,15 @@ private InstitutionPoints calculateInstitutionPoints(Entry<URI, Long> institutio
...
-                   .setScale(SCALE, ROUNDING_MODE);
+        var value = isNullOrZero(institutionCreatorCount) ? ZERO : BigDecimal
+                                                                       .valueOf(institutionCreatorCount)
+                                                                       .divide(new BigDecimal(creatorShareCount),
+                                                                               MATH_CONTEXT);
+        return value.setScale(SCALE, ROUNDING_MODE);
+    }
+
+    private boolean isNullOrZero(Long creatorShareCount) {
+        return creatorShareCount == null || creatorShareCount == 0;
```
Comment (brinxmat) [inline event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/calculator/PointCalculator.java:153]: Objects::nonNull

**Review by brinxmat — APPROVED** [review]
